### PR TITLE
Feat :  Add DataStore storage abstraction with JSON backend (data model layer)

### LINKED ISSUE #1637: Refactor : Storage abstraction layer for environment data
## Is your feature request related to a problem? Please describe.
Visdom's persistence is hard-wired to JSON files on disk. Save, load, delete, layouts, and undo each touch the filesystem directly from server code, so there is no single place that owns "where data lives." That makes it impossible to swap in a different backend, hard to test persistence in isolation, and easy for new code to bypass serialization rules.

## Describe the solution you'd like
A DataStore abstraction that defines the full contract for environment persistence — save, load, delete, list, layouts, and undo history with the existing file-based behaviour implemented as JSONStore behind it. All server routes read and write through the store; nothing touches disk directly. Serialization becomes an implementation detail of the store, not of the request handlers.

## Describe alternatives you've considered
- Jumping straight to a SQL backend: premature. Without the abstraction in place, it just moves the coupling instead of removing it.
- Small helper functions around file I/O: cheaper, but leaves persistence logic scattered and still untestable as a unit.

## Additional context
This is a refactor with no user-visible change; existing environment JSON files stay readable. It unblocks future backends (SQLite/Postgres), search indexing, and proper unit tests for persistence. Best landed as a series of small, reviewable PRs (one concern per PR) rather than a single large change.

### PR DESCRIPTION
**Pull request by Manik-Khajuria-5**
## Description

Adds a small storage abstraction under `py/visdom/data_model/`:

- **`DataStore`** : interface for env persistence (`save_env`, `load_env`, `list_envs`, `delete_env`, ...).
- **`JSONStore`** : default backend; delegates to `serialize_env`, so on-disk format is unchanged.

Definition only, not yet wired into the server. Keeps JSON as default (backward compatible, no new deps); a DB backend can be added later behind the same interface.

## How Has This Been Tested?

- `py/tests/test_data_model.py` : 20 unit tests, hermetic. **20 passed.**

## Types of changes

- [x] New feature (non-breaking)

## Summary by Sourcery

Introduce a pluggable data model layer with a JSON-backed DataStore implementation and corresponding tests.

New Features:
- Add an abstract DataStore interface that defines the storage contract for Visdom environments.
- Provide a JSONStore implementation that persists environments as JSON files with support for hash-based filenames and a no-persistence mode.
- Expose the new DataStore and JSONStore classes via the visdom.data_model package initializer.

Tests:
- Add unit tests covering JSONStore behaviour for both persistent and in-memory configurations, including edge cases around environment listing and hashed IDs.

### CODE REVIEW

**Comment by Manik-Khajuria-5** [comment]
@rajnisht7 @Jayantparashar10 @vedansh-5 @tonypzy please review this It is part of gsoc project and I have implemented it acc to norbert and other mentor guidance

File: py/tests/test_data_model.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+class TestJSONStore(unittest.TestCase):
+    """JSONStore behaviour when persistence is enabled (a real env_path)."""
+
+    def setUp(self):
+        """Give each test a fresh, isolated temp directory as the env_path."""
+        self._tmp = tempfile.TemporaryDirectory()
+        self.env_path = self._tmp.name
+        self.backend = JSONStore(self.env_path)
+
+    def tearDown(self):
```
[the remark this answers is not in the document]
Comment (Manik-Khajuria-5) [reply py/tests/test_data_model.py:29]: done

File: py/tests/test_data_model.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        """save_all persists every environment in state."""
+        state = {"main": _env(), "other": _env()}
+        ret = self.backend.save_all(state)
+        self.assertEqual(sorted(ret), ["main", "other"])
+        self.assertEqual(self.backend.load_env("main"), _env())
+
+    def test_env_named_like_hash_is_not_misread(self):
+        """An env named 'hash_results' is not treated as a hash_<64hex> file."""
+        self.backend.save_env("hash_results", _env())
+        self.assertEqual(self.backend.list_envs(), ["hash_results"])
```
[the remark this answers is not in the document]
Comment (Manik-Khajuria-5) [reply py/tests/test_data_model.py:107]: done

**Review by vedansh-5 — APPROVED** [review]
This was much needed. Changes looks good to me. Nice work

**Review by Jayantparashar10 — APPROVED** [review]

File: py/visdom/data_model/json_store.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    def save_envs(self, state, eids):
+        """Persist the named subset of ``state``; return the ids actually written.
+
+        Each id is sanitised (see :meth:`_safe_eid`) before it becomes a
+        filename, so a crafted id cannot write outside ``env_path``. The real
+        (unsanitised) ids are returned, matching how callers refer to them.
+        """
+        if self.env_path is None:
+            return []
+        written = []
```
[the remark this answers is not in the document]
Comment (Manik-Khajuria-5) [reply py/visdom/data_model/json_store.py:88]: done

**Review by rajnisht7 — APPROVED** [review]

**Review by tonypzy — APPROVED** [review]

**Review by Saksham-Sirohi — APPROVED** [review]
